PhysicsTools/NanoAOD/plugins/NPUTablesProducer.cc

```diff
@@ -24,24 +27,51 @@ class NPUTablesProducer : public edm::global::EDProducer<> {
         void produce(edm::StreamID id, edm::Event& iEvent, const edm::EventSetup& iSetup) const override {
             auto npuTab  = std::make_unique<nanoaod::FlatTable>(1, "Pileup", true);
 
-            edm::Handle<std::vector<PileupSummaryInfo> > npuInfo;
+            
+	    edm::Handle<edm::View<pat::PackedGenParticle> >        genParticles;
+	    double refpvz = -1000.0;
+	    if ( iEvent.getByToken(genTag_, genParticles) ) {
+	      for ( auto genIt = genParticles->begin(); genIt != genParticles->end(); ++genIt ) {
+		if ( genIt->statusFlags().isHardProcess() ) {
+		  refpvz = genIt->vz();
+		  break;
+		}
+	      }
+	    }
+	    
+	    edm::Handle<std::vector<PileupSummaryInfo> > npuInfo;
             if (iEvent.getByToken(npuTag_, npuInfo)) {
-                fillNPUObjectTable(*npuInfo, *npuTab);
+	      fillNPUObjectTable(*npuInfo, *npuTab, refpvz);
             }
 
             iEvent.put(std::move(npuTab));
         }
 
-        void fillNPUObjectTable(const std::vector<PileupSummaryInfo> & npuProd, nanoaod::FlatTable & out) const {
+       void fillNPUObjectTable(const std::vector<PileupSummaryInfo> & npuProd, nanoaod::FlatTable & out, double refpvz) const {
 	  // Get BX 0
 	  unsigned int bx0 = 0;
 	  unsigned int nt = 0;
 	  unsigned int npu = 0;
+
+	  auto zbin = std::lower_bound( vz_.begin(), vz_.end()-1, std::abs(refpvz) );
+	  float pudensity = 0;
+	  float gpudensity = 0;
+	      
 	  for(unsigned int ibx=0; ibx<npuProd.size(); ibx++) {
 	    if(npuProd[ibx].getBunchCrossing()==0) {
 	      bx0 = ibx;
 	      nt = npuProd[ibx].getTrueNumInteractions();
 	      npu = npuProd[ibx].getPU_NumInteractions();
+
+	      std::vector<float> zpositions;
+	      unsigned int nzpositions = npuProd[ibx].getPU_zpositions().size();
+	      for (unsigned int j=0; j<nzpositions; ++j) {
+		zpositions.push_back(npuProd[ibx].getPU_zpositions()[j]);
+		if (std::abs(zpositions.back()- refpvz)<0.1) pudensity++; //N_PU/mm
+		auto bin = std::lower_bound( vz_.begin(), vz_.end()-1, std::abs(zpositions.back()) );
+		if ( bin != vz_.end() && bin==zbin) gpudensity++;
+	      }
+	      gpudensity/=(20.0*( *(zbin+1) - *(zbin) ) ) ;
```

Comment:
This shoukd be:
```suggestion
	      gpudensity/=(20.0*( *(zbin) - *(zbin - 1) ) ) ;
```
because `zbin` refers to the value in `vz_` array that is greater than or equal to `refpvz`. In other words, it's the right edge of the bin. Provided that the binning `zbins` in python config starts from 0 (as you already have), it's safe to decrement the iterator by one step.